models/src/curve_vault.rs

```diff
@@ -0,0 +1,239 @@
+            current_share_price: U256Wrapper::default(),
+            position_count: 0,
+        }
+    }
+}
+
+/// This is a trait that all models must implement.
+impl Model for CurveVault {}
+
+/// This trait works as a contract for all models that need to be upserted into the database.
+#[async_trait]
+impl SimpleCrud<(Option<U256Wrapper>, Option<U256Wrapper>, U256Wrapper)> for CurveVault {
+    /// This method upserts a curve vault into the database.
+    async fn upsert(&self, pool: &PgPool, schema: &str) -> Result<Self, ModelError> {
+        // Generate a unique ID if not already set
+        let id = if self.id == U256Wrapper::default() {
+            // Generate a unique ID based on atom_id/triple_id and curve_number
+            let unique_string = if let Some(atom_id) = &self.atom_id {
+                format!("atom-{}-curve-{}", atom_id, self.curve_number)
+            } else if let Some(triple_id) = &self.triple_id {
+                format!("triple-{}-curve-{}", triple_id, self.curve_number)
+            } else {
+                return Err(ModelError::InsertError("Either atom_id or triple_id must be set".to_string()));
+            };
+            
+            // Hash the unique string to create a numeric ID
+            let mut hasher = DefaultHasher::new();
+            unique_string.hash(&mut hasher);
+            let hash_value = hasher.finish();
+            
+            // Convert the hash to a U256Wrapper by first converting to U256
+            U256Wrapper::from(U256::from(hash_value))
+        } else {
+            self.id.clone()
+        };
+
+        let query = format!(
+            r#"
+            INSERT INTO {}.curve_vault (id, atom_id, triple_id, curve_number, total_shares, current_share_price, position_count)
+            VALUES ($1, $2, $3, $4, $5, $6, $7)
+            ON CONFLICT (id) DO UPDATE SET
+                total_shares = EXCLUDED.total_shares,
+                current_share_price = EXCLUDED.current_share_price,
+                position_count = EXCLUDED.position_count
+            RETURNING id, atom_id, triple_id, curve_number, total_shares, current_share_price, position_count
+            "#,
+            schema,
+        );
+
+        sqlx::query_as::<_, CurveVault>(&query)
+            .bind(id.to_big_decimal()?)
+            .bind(self.atom_id.as_ref().and_then(|w| w.to_big_decimal().ok()))
+            .bind(
+                self.triple_id
+                    .as_ref()
+                    .and_then(|w| w.to_big_decimal().ok()),
+            )
+            .bind(self.curve_number.to_big_decimal()?)
+            .bind(self.total_shares.to_big_decimal()?)
+            .bind(self.current_share_price.to_big_decimal()?)
+            .bind(self.position_count)
+            .fetch_one(pool)
+            .await
+            .map_err(|e| ModelError::InsertError(e.to_string()))
+    }
+
+    /// Finds a curve vault by its composite key (atom_id, triple_id, curve_number).
```

Comment:
We probably also want a method to get the composite key by taking in atom_id, triple_id, curve_id